### PR TITLE
AcraTranslator refactoring: service start up and shutdown, signals listener

### PR DESCRIPTION
**Pull request by storojs72**
This PR introduces several updates for our AcraTranslator service. 

In general I refactored main component (`ReaderServer`) and added support of SIGHUP handling to the service. What is refactored: 
- Added new `StartFromFileDescriptor` function that allows to run HTTP and gRPC servers
having listener obtained from correspondent descriptor, instead of creating a new ones in order to preserve
current listeners in new process;
- Splitted monster `Start` function on lower components `startHTTP`, `startGRPC` in order
to improve readability;
- Wrapped `Start` and `StartFromFileDescriptor` with timeout exit (single point) and avoiding
hidden `os.Exit` calls. This is extremely important for EE version that relies on CE ReaderServer
component;
- Moved from 'multiply listeners - multiply connections' to 'single listener - multiply connection'
with transparent management of stopping/closing it. This is also important for better controlling of
code execution;
- Added some getters which is needed by EE version.

There are two questions about previous implementation that were not obvious for me (most likely they are addressed to @Lagovas), and probably we can discuss it while review process: 1) why previously we had 'multiply listeners - multiply connection' model (when we called `HandleClientConnection`, it was a `AcceptConnection` function inside that created internally new listener and returned accepted connection via channel, their closing was managed via child contexts cancelling in separate goroutine) if `listener` itself allows calling `Accept` function safely from multiply goroutines; and 2) why does switching to new process works, if file descriptors created while SIGHUP processing and on service starting are different? (I asked @Lagovas about it and we came to the point that we know order number of Acra descriptor - it's 3, but it seems that we need not the order number but correct actual value of `fd` to start)

How did I test this? Manually. I created a following methodology for testing it (locally run AcraTranslator, AcraConnector and 2 toy applications: one for sending HTTP and second - for gRPC requests):

1) Test HTTP component without/with active connections;
2) Test gRPC component without/with active connections;
3) Test HTTP + gRPC components that run simultaneously without/with active connections;

For sending HTTP requests I used the following piece of code (thanks to our documentation - but some examples are outdated there):
```
for run in {1..100}; 
do curl -X POST --data-binary @client.acrastruct --header "Content-Type: application/octet-stream" http://127.0.0.1:8000/v1/decrypt; 
done
```

For sending gRPC requests I used following go code:
```
conn, err := grpc.Dial("127.0.0.1:8001", grpc.WithInsecure())
	if err != nil {
		t.Fatal(err)
	}
	defer conn.Close()


	client := grpc1.NewReaderClient(conn)

	for start := time.Now(); time.Since(start) < time.Second * 10; {
		_, _ = client.Decrypt(context.Background(),
			&grpc1.DecryptRequest{
				ClientId:   []byte("testclientid"),
				ZoneId:     []byte("test_zone"),
				Acrastruct: []byte("blabla"),
			},
			grpc.EmptyCallOption{})
	}
```
While those applications were running, I sent (multiply times) SIGHUP signal to the AcraTranslator and observed that service restarted and there were no cut connections. Testing without connections are rather straightforward.

### CODE REVIEW

**Comment by shadinua** [comment]
> For sending HTTP requests I used the following piece of code (thanks to our documentation - but some examples are outdated there):
> 
> ```
> for run in {1..100}; 
> do curl -X POST --data-binary @client.acrastruct --header "Content-Type: application/octet-stream" http://127.0.0.1:8000/v1/decrypt; 
> done
>```

In the example above we send requiests sequentially. I'd suggest to make a simple, but effective impromevent:
```bash
TEST_THREADS=20
TEST_CYCLES=100
for i in $(seq 1 $TEST_THREADS); do
    for j in $(seq 1 $TEST_CYCLES); do
        curl -X POST \
            --data-binary @client.acrastruct \
            --header "Content-Type: application/octet-stream" \
            http://127.0.0.1:8000/v1/decrypt
    done &
done
```

File: cmd/acra-translator/acra-translator.go
Comment (ilammy) [inline cmd/acra-translator/acra-translator.go:184]: Do I understand it correctly that if this fails then we still go on with restarting AcraTranslator? It will not reuse the listening sockets and will try to reopen them. It will probably not work since this process is still active, but I guess we give it a chance? Why not just fail immediately like we do with all errors that follow?

I guess we can't bail out at this point since `readerServer` has been already stopped, so our best bet is to die with an error and rely on service manager to restart AcraTranslator.

(Also, I believe the log message should use `GracefulEnv` constant value, not just `gracefulEnv` string.)

File: cmd/acra-translator/acra-translator.go
Comment (ilammy) [inline cmd/acra-translator/acra-translator.go:188]: You can use `listener` variable inside the if clause instead of getting it again.

File: cmd/acra-translator/acra-translator.go
Comment (ilammy) [inline cmd/acra-translator/acra-translator.go:268]: I don't quite understand what happens here. Could you please elaborate?

What I find strange is that zero is a valid file descriptor value (usually stdin), I don't think it's okay to use it like this.

File: network/utils.go
Comment (ilammy) [inline network/utils.go:111]: I'd suggest to extract the error into

```go
var ErrNilListener = errors.New("nil listener")
```

File: cmd/acra-translator/common/network.go
Comment (ilammy) [inline cmd/acra-translator/common/network.go:36]: So... If understand correctly, previously closing the listener resulted in an `accept` error being sent into `errCh`, and now we avoid that. However, the context is basically useless now. Well... This is a weird flex. I don't see any immediate way to improve it, but I don't like it being counterintuitive: usually methods which accept a context can be expected to be cancelable via that context. This is not the case here.

File: cmd/acra-translator/server/server.go
Comment (ilammy) [inline cmd/acra-translator/server/server.go:79]: I don't see how this renaming improves readability very much, but okay...

**Review by ilammy — COMMENTED** [review]
I have yet to read `cmd/acra-translator/server/server.go` properly. Here are some drive-by comments.

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [reply cmd/acra-translator/acra-translator.go:188]: Agree, I'd suggest to use existing `listener`.

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:199]: What is "acra-socket"?
I mean from user perspective. What users should think about when they read "failed to get acra-socket file descriptor ..". Maybe use just "failed to get Acra file descriptor .."?

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:219]:
```suggestion
			log.Warningf("Server shutdown timeout: %d active connections will be cut", readerServer.GetConnectionManager().Counter)
```

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [inline cmd/acra-translator/acra-translator.go:233]: 👍

File: cmd/acra-translator/common/network.go
Comment (vixentael) [inline cmd/acra-translator/common/network.go:45]: Make this log Debug level?
I don't think that end users should see this log.

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:224]: copy paste detected :)
```suggestion
				Errorln("Can't create gRPC listener from specified connection string")
```

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:223]:
```suggestion
			log.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorTranslatorCantAcceptNewGRPCConnection).
```

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:271]:
```suggestion
func (server *ReaderServer) startGRPC(logger *log.Entry, decryptorData *common.TranslatorData, errCh chan<- error, listener net.Listener) {
```

I see that functions and variables in code use HTTP and GRPC as ALL CAPS format, let's continue doing so.

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:295]:
```suggestion
			logger.WithError(err).Errorln("Can't create new gRPC server")
```

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:348]:
```suggestion
			logger.Errorln("Can't create new file from descriptor for Acra HTTP listener")
```

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:373]:
```suggestion
			logger.Errorln("Can't create new file from descriptor for Acra gRPC listener")
```

File: cmd/acra-translator/server/server.go
Comment (vixentael) [inline cmd/acra-translator/server/server.go:468]:
```suggestion
			log.WithError(err).Warningln("Error occured while stopping gRPC listener")
```

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [reply cmd/acra-translator/acra-translator.go:183]: or, as @ilammy mentioned, log value of `GracefulRestartEnv` constant direcly

File: cmd/acra-translator/server/server.go
Comment (ilammy) [inline cmd/acra-translator/server/server.go:130]: Doesn't `errors.As` assign to its second argument if the cast is successful? `context.Canceled` is global "constant variable" from the `context` package. Maybe you meant to use `errors.Is` here instead?

There are other places where `errors.As` is used similarly and may need attention.

File: cmd/acra-translator/server/server.go
Comment (ilammy) [inline cmd/acra-translator/server/server.go:237]: I'd actually log the error here again with some high-level information, indicating that this error has caused the connection to close. The goroutines worked so hard to put this error into the channel, it would be a waste to not use it.

Ditto in `StartFromFileDescriptor`.

File: cmd/acra-translator/server/server.go
Comment (ilammy) [inline cmd/acra-translator/server/server.go:324]: I'd call it something like `waitForExitTimeout` to avoid making an impression that calling this method *causes* the server to exit.

File: cmd/acra-translator/server/server.go
Comment (ilammy) [inline cmd/acra-translator/server/server.go:371]: While we're here, maybe the `"/tmp/acra-translator_grpc"` should be extracted into some constant. This value is actually irrelevant, it's only a placeholder. But would be nice to indicate to the reader that it's not some random value. `grpcFilenamePlaceholder`? Does anyone have better ideas?

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:50]: Definitely. But can you please explain once again, why it works if when we fork process (it seems to me) we need _not an order number of descriptor_ but _descriptor value_ to pass, isn't it?

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:134]: Do you mean that we need an ability to cancel actual connection in such a way? Does it relate to the previous usage of multiply listeners that create connections instead of one listener as currently?

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:184]: Yes. It will use `Start` function without using descriptors. And actually I don't know. I ported it from our AcraServer (https://github.com/cossacklabs/acra/blob/master/cmd/acra-server/acra-server.go#L391). cc @Lagovas

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:188]: That's true. Thank you

File: network/utils.go
Comment (storojs72) [reply network/utils.go:111]: Definitely. Thank you

File: cmd/acra-translator/common/network.go
Comment (storojs72) [reply cmd/acra-translator/common/network.go:36]: This is an afford to handle a case of `"use of closed network connection"` error. Yes, context is almost useless, except that we extract logger from it (which is also not obvious for me - why)

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:49]: Ok

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:183]: Ok. Probably you meant `log.WithError(err).Errorf("Unexpected error on os.Setenv, graceful restart won't work. Please check env variables, especially %s", GracefulRestartEnv)`

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:199]: legacy from `AcraServer`. Yes `failed to get Acra gRPC file descriptor` is really better

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:271]: ok

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:295]: ok, thank you

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:348]: ok

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:373]: ok, thank you

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:468]: ok

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:130]: TBH, I didn't think too much on it. Now after reading descriptions of `Is` and `As` it seems to me that `As` may panic if it can't cast and `Is` will return false in such case, so `Is` is really more appropriate here.

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:237]: ok

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:324]: ok, thank you

File: cmd/acra-translator/server/server.go
Comment (storojs72) [reply cmd/acra-translator/server/server.go:371]: Yes. It would be better. Thank you

**Comment by storojs72** [comment]
@shadinua thank you for extension of the HTTP app. I will use your version for further testing

File: cmd/acra-translator/acra-translator.go
Comment (vixentael) [reply cmd/acra-translator/acra-translator.go]: yes, looks good

File: cmd/acra-translator/acra-translator.go
Comment (Lagovas) [reply cmd/acra-translator/acra-translator.go:274]: we should check it in function [StartFromFileDescriptor](https://github.com/cossacklabs/acra/blob/f8ab52028858926fd2ed645dd092a371cea62db5/cmd/acra-translator/server/server.go#L344). Add some value which will represent null as integer value

File: cmd/acra-translator/acra-translator.go
Comment (Lagovas) [reply cmd/acra-translator/acra-translator.go]: Okay. When we do fork, child process inherit parent's descriptors. And probably all of them will be closed in parent process on exit if they haven't specific flag to not close on exit. Golang has wrapper around fork. It has function which do some preparation before return control, and they duplicate all descriptors which you can specify directly to be sure that they will not be closed.
So, our goal to fork new process and pass descriptors of our current listeners. For example server listener has fd=23 and api listeners has fd=152. So we should tell our child process that it should save and reuse these descriptors and use them as listeners. But every new process has own descriptor table and we will not know which of them responsible for these listeners. So we pass them directly, golang duplicate them and it use first free descriptor values. 3 and 4 (probably previous author of these code checked it and verified that it always use these values:) and we mark them as responsible for server and api listeners, and reuse them.

Comment (storojs72) [reply cmd/acra-translator/acra-translator.go]: Thank you

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go]: ok. Will change on `Fatalf`

File: cmd/acra-translator/acra-translator.go
Comment (storojs72) [reply cmd/acra-translator/acra-translator.go:274]: Yes. That is weird piece of code that I wrote while debugging. Now I think we can do just `readerServer.StartFromFileDescriptor(mainContext, DescriptorHTTP, DescriptorGRPC)` in any case. Inside `StartFromFileDescriptor` we have `http`, `grpc` conditions that will use appropriate descriptors

File: cmd/acra-translator/common/network.go
Comment (storojs72) [reply cmd/acra-translator/common/network.go:36]: @ilammy, now I can't reproduce the `use of closed network connection` error with the EE translator. Probably it was due to some intermediate version of code while debugging, so now I think we can avoid it and use just:
```
conn, err := listener.Accept()
if err != nil {
	logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCantAcceptNewConnections).WithError(err).Errorln("Error on accept connection")
	errCh <- err
	return
}
```
which is much more intuitive

**Comment by storojs72** [comment]
@ilammy, @Lagovas, @vixentael after cleaning the source code from weird leftovers after debugging, it seems to be good to merge. Only SSL issue is left to be resolved

**Review by ilammy — APPROVED** [review]
Well, my comments have all been addressed so here's my green thingie :)

There are still a couple of spelling nits from @vixentael though.

> Only SSL issue is left to be resolved for go 1.15.2 in order to pass CircleCI checks

I'm working on it. The issue has been isolated and fixed. I'm doing some final tests and preparing the PR. 

@storojs72, if you wish to merge this PR, you can go ahead and drop 1.15 from tested versions. Or you can wait for that PR to be merged.

**Review by Lagovas — APPROVED** [review]
lgtm, but let's merge after fixing @vixentael comments)

**Comment by storojs72** [comment]
@Lagovas , @ilammy can You please point, what should I actually fix, because it seems that I added those changes in https://github.com/cossacklabs/acra/pull/414/commits/3673fcf368ce054660393f5b01922ce0757c999d isn't it?

**Comment by ilammy** [comment]
@storojs72, I've applied the easy ones, changing log lines. It seems that only this one remains: https://github.com/cossacklabs/acra/pull/414#discussion_r491192768
I'm not sure that `EventCodeErrorTranslatorCantAcceptNewGRPCConnection` exists (don't see it it `logging/event_codes.go`), you may need to add it first before updating the log.

**Comment by ilammy** [comment]
Welp, it *does* conflict with recently merged fix for TLS. Okay, let me resolve the merge conflict, and I'll apply the remaining fix suggested by @vixentael along the way.

**Comment by ilammy** [comment]
This should do it. @vixentael, please confirm.

@storojs72, I'm sorry for barging into your branch. Please do `git pull` first if you plan to add something else.

**Review by vixentael — APPROVED** [review]
great work, thank you everybody!

**Comment by storojs72** [comment]
Nothing more. Let's merge it

**Pull request merged**
